src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,308 @@
+
+```
+  +----+----+----+----+----+----+----+----+----+----+
+  | 47 | s  | e  | r  | v  | i  | c  | e  | C0 | 00 |
+  +----+----+----+----+----+----+----+----+----+----+
+```
+
+- `0x47` (`0b0100_0111`) - label starting with `_` with `len=7`.
+- `0xC0` (`0b110_0000`) - Commonly used constant label - code 0 for `_udp`.
+- `0x00` end.
+
+`2906C908D115D362-8FC7772401CD0696` (matter service instance label example) encoded using `17` bytes (vs uncompressed DNS encoding of `34` bytes).
+
+- `0xE1` (`0b1110_0001`) - Commonly used generative patterns (code 1)
+- `0x29` `0x06` `0xC9` `0x08` `0xD1` `0x15` `0xD3` `0x62` (first 64-bit value)
+- `0x8F` `0xC7` `0x77` `0x24` `0x01` `0xCD` `0x06` `0x96` (second 64-bit value)
+
+## Message Format
+
+- [Header Block](#header-block)
+- Zero or more
+  - [Add Service Block](#add-service-block) or [Remove Service Block](#remove-service-block)
+- [Host Block](#host-block)
+- [Footer Block](#footer-block)
+
+## Header Block
+
+### Header Dispatch
+
+```
+    0      1     2     3     4     5     6     7
+  +-----+-----+-----+-----+-----+-----+-----+-----+
+  |  0  |  1  |  1  |  0  |  1  |  1  |  Z  |  T  |
+  +-----+-----+-----+-----+-----+-----+-----+-----+
+```
+
+- The initial bits in dispatch byte are fixed to differentiate it from the general DNS header used in SRP updates.
+- `Z` flag (Zone)
+  - `0`: Zone name is elided - use the default `default.service.arpa`.
+  - `1`: Zone name is encoded within the message.
+- `T` flag (Default TTL)
+  - `0`: Default TTL is elided - use `7200` (2 hours).
+  - `1`: Default TTL is encoded in the message using compact integer format.
+
+### Header Block Format
+
+- Message ID: Two bytes, big-endian.
+- Dispatch byte.
+- Zone (domain) name: Included if not elided (indicated by the `Z` flag).
+- Default TTL value: Encoded using compact integer format if not elided (indicated by the `T` flag).
+- Host labels: Excludes the domain name and ends with an empty label.
+
+#### Distinguishing from SRP Update Message
+
+DNS header follows the format below (RFC1035):
+
+```
+   0  1  2  3  4  5  6  7  8  9  0  1  2  3  4  5
+  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+  |                    ID                         |
+  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+  |QR|   Opcode  |AA|TC|RD|RA|   Z    |   RCODE   |
+  +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+
+```
+
+The first 3 bytes can be used to differentiate a coded message from an SRP update message or other DNS message types. In the coded message, the message ID is encoded in the first two bytes, similar to a general DNS header. The fixed initial bit pattern in the header dispatch byte, when mapped to the DNS header, correspond to `QR = 0`, `Opcode = 0b1101 = 13`, and `AA = 1`. The combination of `OpCode = 13` (currently unused/unassigned) and `AA = 1` is used to distinguish the coded message. An SRP Update message uses the `Opcode = 5`.
```

Comment:
This will not fly in the IETF. But no worries, check out RFC2136, which defines these bits for DNS updates specifically. We can just use Opcode=5 as usual and set bit 5, which is not AA in a DNS Update, to 1.

Reply:
Sounds good. Updated in new push to use `Opcode=5` with bit `5` set.